tds/src/main/java/thredds/server/wms/ThreddsWmsServlet.java

```diff
@@ -95,60 +96,75 @@ protected void dispatchWmsRequest(String request, RequestParams params, HttpServ
+    }
+
+    final CachedWmsCatalogue cachedWmsCatalogue = catalogueCache.getIfPresent(tdsDatasetPath);
+    final long lastModified = ncd.getLastModified();
+
+    if (cachedWmsCatalogue != null && cachedWmsCatalogue.lastModified >= lastModified) {
+      // Must update NetcdfDataset to ensure file resources are reacquired, as this has been closed.
+      // But we don't need to recreate the ThreddsWmsCatalogue as it is up-to-date according to the last modified
+      cachedWmsCatalogue.wmsCatalogue.setNetcdfDataset(ncd);
+      return cachedWmsCatalogue.wmsCatalogue;
+    } else {
+      // Create and put/ replace in cache
+      ThreddsWmsCatalogue threddsWmsCatalogue = new ThreddsWmsCatalogue(ncd, tdsDatasetPath);
+      catalogueCache.put(tdsDatasetPath, new CachedWmsCatalogue(threddsWmsCatalogue, lastModified));
+      cacheLoads++;
+      return threddsWmsCatalogue;
+    }
   }
 
-  // package private for testing
-  static boolean useCachedCatalogue(String tdsDatasetPath) {
-    if (containsCachedCatalogue(tdsDatasetPath)) {
-      // This date last modified will be updated e.g. in the case of an aggregation with a recheckEvery
-      final long netcdfDatasetLastModified = catalogueCache.get(tdsDatasetPath).wmsCatalogue.getLastModified();
-      final long cacheLastModified = catalogueCache.get(tdsDatasetPath).lastModified;
-      return cacheLastModified >= netcdfDatasetLastModified;
+  private static NetcdfDataset acquireNetcdfDataset(HttpServletRequest httpServletRequest,
+      HttpServletResponse httpServletResponse, String tdsDatasetPath) throws IOException {
+    NetcdfFile ncf = TdsRequestedDataset.getNetcdfFile(httpServletRequest, httpServletResponse, tdsDatasetPath);
+    if (TdsRequestedDataset.useNetcdfJavaBuilders()) {
+      return NetcdfDatasets.enhance(ncf, NetcdfDataset.getDefaultEnhanceMode(), null);
+    } else {
+      return NetcdfDataset.wrap(ncf, NetcdfDataset.getDefaultEnhanceMode());
+    }
+  }
+
+  public static void showCache(Formatter formatter) {
+    formatter.format("%nWmsCache:%n");
+    formatter.format("numberOfEntries=%d, ", getNumberOfEntries());
+    formatter.format("loads=%d, ", getCacheLoads());
+    formatter.format("evictionCount=%d ", catalogueCache.stats().evictionCount());
+    formatter.format("%nentries:%n");
+    for (Map.Entry<String, CachedWmsCatalogue> entry : catalogueCache.asMap().entrySet()) {
+      formatter.format("  %s%n", entry.getKey());
     }
-    return false;
+  }
+
+  public static void resetCache() {
+    catalogueCache.invalidateAll();
+    cacheLoads = 0;
   }
 
   // package private for testing
   static boolean containsCachedCatalogue(String tdsDatasetPath) {
-    return catalogueCache.containsKey(tdsDatasetPath);
+    return catalogueCache.asMap().containsKey(tdsDatasetPath);
+  }
+
+  static long getNumberOfEntries() {
+    return catalogueCache.size();
+  }
+
+  static int getCacheLoads() {
+    return cacheLoads;
```

Comment:
Are we using the number of cache loads somewhere/need to track it?

Reply:
just in the `TestWmsCache` to test if it reloaded vs reused the cached dataset